br/pkg/backup/prepare_snap/prepare.go

```diff
@@ -0,0 +1,430 @@
+				return errors.Annotatef(err, "failed to finalize the prepare stream for %d", id)
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		logutil.CL(ctx).Warn("failed to finalize some prepare streams.", logutil.ShortError(err))
+		return err
+	}
+	logutil.CL(ctx).Info("all connections to store have shuted down.")
+	for {
+		select {
+		case event := <-p.eventChan:
+			if err := p.onEvent(ctx, event); err != nil {
+				return err
+			}
+		default:
+			return nil
+		}
+	}
+}
+
+func (p *Preparer) batchEvents(evts *[]event) {
+	for {
+		select {
+		case evt := <-p.eventChan:
+			*evts = append(*evts, evt)
+		default:
+			return
+		}
+	}
+}
+
+// WaitAndHandleNextEvent is exported for test usage.
+// This waits the next event (wait apply done, errors, etc..) of preparing.
+// Generally `DriveLoopAndWaitPrepare` is all you need.
+func (p *Preparer) WaitAndHandleNextEvent(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		logutil.CL(ctx).Warn("User canceled.", logutil.ShortError(ctx.Err()))
+		return ctx.Err()
+	case evt := <-p.eventChan:
+		logutil.CL(ctx).Debug("received event", zap.Stringer("event", evt))
+		events := []event{evt}
+		p.batchEvents(&events)
+		for _, evt := range events {
+			err := p.onEvent(ctx, evt)
+			if err != nil {
+				return errors.Annotatef(err, "failed to handle event %v", evt)
+			}
+		}
+		return p.MaybeFinish(ctx)
+	case <-p.retryChan():
+		return p.workOnPendingRanges(ctx)
+	}
+}
+
+func (p *Preparer) removePendingRequest(r *metapb.Region) bool {
+	r2, ok := p.inflightReqs[r.GetId()]
+	if !ok {
+		return false
+	}
+	matches := r2.GetRegionEpoch().GetVersion() == r.GetRegionEpoch().GetVersion() &&
+		r2.GetRegionEpoch().GetConfVer() == r.GetRegionEpoch().GetConfVer()
+	if !matches {
+		return false
+	}
+	delete(p.inflightReqs, r.GetId())
+	return true
+}
+
+func (p *Preparer) onEvent(ctx context.Context, e event) error {
+	switch e.ty {
+	case eventMiscErr:
+		// Note: some of errors might be able to be retry.
+		// But for now it seems there isn't one.
+		return errors.Annotatef(e.err, "unrecoverable error at store %d", e.storeID)
+	case eventWaitApplyDone:
+		if !p.removePendingRequest(e.region) {
+			logutil.CL(ctx).Warn("received unmatched response, perhaps stale, drop it", zap.Stringer("region", e.region))
+			return nil
+		}
+		r := rangeOrRegion{
+			id:       e.region.GetId(),
+			startKey: e.region.GetStartKey(),
+			endKey:   e.region.GetEndKey(),
+		}
+		if e.err != nil {
+			logutil.CL(ctx).Warn("requesting a region failed.", zap.Uint64("store", e.storeID), logutil.ShortError(e.err))
+			p.failed = append(p.failed, r)
+			if p.nextRetry != nil {
+				p.nextRetry.Stop()
+			}
+			// Reset the timer so we can collect more regions.
+			// Note: perhaps it is better to make a deadline heap or something
+			// so every region backoffs the same time.
+			p.nextRetry = time.NewTimer(p.RetryBackoff)
+			return nil
+		}
+		if item, ok := p.waitApplyDoneRegions.ReplaceOrInsert(r); ok {
```

Comment:
when overlapped happened. Do we need set overlapped range to fail?

Reply:
If that range is larger, it may be retry after we checking the holes in `AdvanceStep`, if it is smaller, I think it can be ignored.